### PR TITLE
Warn about creating multiple replicas from kubernetes YAML

### PR DESCRIPTION
**Pull request by dcermak**
See: https://github.com/containers/podman/pull/16766#issuecomment-1378696111

```release-note
- Limit replicas to one when creating pods/containers from kubernetes YAML
```

This fixes: https://github.com/containers/podman/issues/16765
cc @vrothberg

### CODE REVIEW

File: pkg/domain/infra/abi/play.go [hunk 1]
```diff
@@ -1255,10 +1254,12 @@ func (ic *ContainerEngine) PlayKubeDown(ctx context.Context, body io.Reader, opt
 			if deploymentYAML.Spec.Replicas != nil {
 				numReplicas = *deploymentYAML.Spec.Replicas
 			}
-			for i := 0; i < int(numReplicas); i++ {
-				podName := fmt.Sprintf("%s-pod-%d", deploymentName, i)
-				podNames = append(podNames, podName)
+			if numReplicas > 1 {
```
Comment (ygalblum) [inline pkg/domain/infra/abi/play.go:1257]: This code is in `PlayKubeDown` so I'm not sure the warning is needed. Having said that, if we want to print something, we should rephrase to indicate `down` instead of `creating...`

**Comment by rhatdan** [comment]
Replicas make little sense on a single node, so warning is the best we can do.

**Comment by rhatdan** [comment]
I don't think we really supported Replicas before.

**Comment by mheon** [comment]
We did create them (though I agree they didn't really function properly - or at all, really)

File: pkg/domain/infra/abi/play.go [hunk 2]
```diff
@@ -345,20 +344,20 @@ func (ic *ContainerEngine) playKubeDeployment(ctx context.Context, deploymentYAM
 	if deploymentYAML.Spec.Replicas != nil {
 		numReplicas = *deploymentYAML.Spec.Replicas
 	}
+	if numReplicas > 1 {
+		logrus.Warnf("Limiting replica count to 1, more than one replica is not supported in podman")
```
Comment (TomSweeneyRedHat) [inline pkg/domain/infra/abi/play.go:348]:
```suggestion
		logrus.Warnf("Limiting replica count to 1, more than one replica is not supported by Podman")
```

File: pkg/domain/infra/abi/play.go [hunk 3]
```diff
@@ -1255,10 +1254,12 @@ func (ic *ContainerEngine) PlayKubeDown(ctx context.Context, body io.Reader, opt
 			if deploymentYAML.Spec.Replicas != nil {
 				numReplicas = *deploymentYAML.Spec.Replicas
 			}
-			for i := 0; i < int(numReplicas); i++ {
-				podName := fmt.Sprintf("%s-pod-%d", deploymentName, i)
-				podNames = append(podNames, podName)
+			if numReplicas > 1 {
+				logrus.Warnf("Limiting replica count to 1, more than one replica is not supported in podman")
```
Comment (TomSweeneyRedHat) [inline pkg/domain/infra/abi/play.go:1258]: if you take the "by Podman" above, please do the same here.

**Comment by TomSweeneyRedHat** [comment]
If we want to do just a warning, the approach LGTM.  I'll let others debate the merit of that approach.

**Review by ygalblum — COMMENTED** [review]
Code LGTM, but, need to see why tests are failing. Maybe there are some more tests that need adjustment

**Comment by dcermak** [comment]
@ygalblum Tests should be green now.

File: test/e2e/play_kube_test.go [hunk 4]
```diff
@@ -4724,6 +4712,7 @@ cgroups="disabled"`), 0644)
 	})
 
 	It("podman kube --quiet with error", func() {
+		SkipIfNotRootless("We need to create an error trying to bind to port 80")
```
Comment (vrothberg) [inline test/e2e/play_kube_test.go:4715]: This seems out of context.  Please move it to another commit (with explanation) if needed.

**Review by vrothberg — COMMENTED** [review]
Need a test that expects the warning to be printed to make sure we don't silently regress in the future.

File: test/e2e/play_kube_test.go (hunk 4 above)
Comment (dcermak) [reply test/e2e/play_kube_test.go:4715]: So the issue here is, that this new test was relying on that playing this kubernetes yaml would create an error as it would bind two pods (due to replicas = 2) to the same port. With my PR, that would no longer error out and this test would now start failing.

I have added it to the same commit, so that every commit passes the CI and can thus be safely backported. I can move it to another commit though, if you prefer that.

Comment (vrothberg) [reply test/e2e/play_kube_test.go:4715]: Ah, OK.  That makes sense, thank you!  No need to change anything.  Was a brain fart on my end.

Comment (dcermak) [reply test/e2e/play_kube_test.go:4715]: That hardly qualifies as a brain fart…

**Comment by dcermak** [comment]
@vrothberg I have added a test for the log output and also tried to add a test that verifies that podman really only creates a single pod. But I'm not really happy with it, because it relies on podman's stdout. However, I did not want to use the `getPodNameInDeployment` helpers, as these use internal knowledge about podman's naming scheme and will by design never give you more than one pod name.

**Review by vrothberg — APPROVED** [review]
LGTM, thank you @dcermak !

**Review by ygalblum — COMMENTED** [review]
LGTM

**Comment by vrothberg** [comment]
Tests aren´t happy yet

**Comment by dcermak** [comment]
> Tests aren´t happy yet

Yes, the test checking for the log warning was still running for remote integration tests, but iirc that is not supported. Should be fixed now :crossed_fingers:

**Comment by dcermak** [comment]
> Tests aren´t happy yet

After a few complimentary force-pushes, they were convinced to be happy ;-)

**Comment by TomSweeneyRedHat** [comment]
LGTM
and happy green test buttons.

**Comment by thedanielenavarra** [comment]
Hi everyone, 
First of all I'd like to thank you guys for the amazing work on Podman
Here's my question: since the [official docs](https://docs.podman.io/en/latest/markdown/podman-kube-play.1.html)  indicate the **_replica_** field as supported:
podman-replicas

Is it going to be dropped as a supported feature or will it be **actually** implemented?
In my humble opinion the concept of replicas is not secondary,  it would be very useful to have them

**Comment by thedanielenavarra** [comment]
Hi everyone, 
First of all I'd like to thank you guys for the amazing work on Podman
Here's my question: since the [official docs](https://docs.podman.io/en/latest/markdown/podman-kube-play.1.html)  indicate the **_replica_** field as supported:
podman-replicas

Is it going to be dropped as a supported feature or will it be **actually** implemented?

Since Podman can be used to test YAML files that will eventually be applied on Kubernetes clusters, in my humble opinion the concept of replicas is not secondary,  it would be very useful to test them even on a single node.

Please let me know if this is the wrong place to ask this question :smile: 
Thanks in advance

**Comment by dcermak** [comment]
> Is it going to be dropped as a supported feature or will it be actually implemented?

I think we have effectively dropped support for this feature with this PR, as we deemed it to not really make a lot of sense to offer multiple replicas on a single host without built-in support for an ingress.

**Comment by ygalblum** [comment]
> I think we have effectively dropped support for this feature with this PR, as we deemed it to not really make a lot of sense to offer multiple replicas on a single host without built-in support for an ingress.

Yes we have. But, Podman does not fail when replica is set to more than 1, it just logs a warning. So, your YAML file will continue to work. Just won't spin multiple replicas
